catalyst/callbacks/lambda_preprocess.py

```diff
@@ -0,0 +1,158 @@
+    ):
+        """Wraps input for your callback with specified function.
+
+        Args:
+            lambda_fn (Callable): Function to apply.
+            keys_to_apply (Union[List[str], str], optional): Keys in batch dict to apply function.
+                Defaults to ["s_hidden_states", "t_hidden_states"].
+            output_keys (Union[List[str], str], optional): Keys for output.
+                If None then will apply function inplace to ``keys_to_apply``.
+                Defaults to None.
+
+        Raises:
+            TypeError: When keys_to_apply is not str or list.
+        """
+        super().__init__(order=CallbackOrder.Internal)
+        if not isinstance(keys_to_apply, (list, str)):
+            raise TypeError("keys to apply should be str or list of str.")
+        if output_keys is not None:
+            if not isinstance(output_keys, (list, str)):
+                raise TypeError("output keys should be str or list of str.")
+        self.keys_to_apply = keys_to_apply
+        self.output_keys = output_keys
+        self.lambda_fn = lambda_fn
+
+    def on_batch_end(self, runner) -> None:
+        """
+        On batch end action.
+
+        Args:
+            runner: runner for the experiment.
+
+        Raises:
+            TypeError: If lambda_fn output has a wrong type.
+
+        """
+        batch = runner.batch
+
+        if isinstance(self.keys_to_apply, list):
+            fn_inp = [batch[key] for key in self.keys_to_apply]
+            fn_output = self.lambda_fn(*fn_inp)
+            if isinstance(fn_output, tuple):
+                if self.output_keys is not None:
+                    if not isinstance(self.output_keys, list):
+                        raise TypeError(
+                            "Unexpected output from function. "
+                            "For output key type string expected one element, got tuple."
+                        )
+                    iter_keys = self.output_keys
+                else:
+                    iter_keys = self.keys_to_apply
+                for idx, key in enumerate(iter_keys):
+                    batch[key] = fn_output[idx]
+            elif isinstance(fn_output, dict):
+                for outp_k, outp_v in fn_output.items():
+                    batch[outp_k] = outp_v
+            else:
+                if self.output_keys is not None:
+                    if not isinstance(self.output_keys, str):
+                        raise TypeError(
+                            "Unexpected output from function. "
+                            "For output key type List[str] expected tuple, got one element."
+                        )
+                    output_key = self.output_keys
+                else:
+                    output_key = self.keys_to_apply
+                batch[output_key] = fn_output
+        elif isinstance(self.keys_to_apply, str):
+            batch[self.keys_to_apply] = self.lambda_fn(self.keys_to_apply)
```

Comment:
let's move the logic definition of batch-handling to the `__init__` so it would be much faster
(if-else operation at each batch - looks too slow ;) )